internal/crd/demand_informer.go

```diff
@@ -0,0 +1,142 @@
+	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
+	"github.com/palantir/witchcraft-go-logging/wlog/wapp"
+	apiextensionsclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	clientcache "k8s.io/client-go/tools/cache"
+)
+
+const (
+	informerSyncRetryCount          = 5
+	informerSyncTimeout             = 2 * time.Second
+	informerSyncRetryInitialBackoff = 500 * time.Millisecond
+)
+
+// LazyDemandInformer checks for Demand CRD existence and creates a
+// demand informer if it exists.
+type LazyDemandInformer struct {
+	informerFactory     ssinformers.SharedInformerFactory
+	apiExtensionsClient apiextensionsclientset.Interface
+	ready               chan struct{}
+	informer            v1alpha1.DemandInformer
+	lock                sync.RWMutex
+}
+
+// NewLazyDemandInformer constructs a new LazyDemandInformer instance
+func NewLazyDemandInformer(
+	informerFactory ssinformers.SharedInformerFactory,
+	apiExtensionsClient apiextensionsclientset.Interface) *LazyDemandInformer {
+	return &LazyDemandInformer{
+		informerFactory:     informerFactory,
+		apiExtensionsClient: apiExtensionsClient,
+		ready:               make(chan struct{}),
+	}
+}
+
+// Informer returns the informer instance if it is initialized, returns false otherwise
+func (ldi *LazyDemandInformer) Informer() (v1alpha1.DemandInformer, bool) {
+	ldi.lock.RLock()
+	defer ldi.lock.RUnlock()
+	select {
+	case <-ldi.Ready():
+		return ldi.informer, true
+	default:
+		return nil, false
+	}
+}
+
+// Ready returns a channel that will be closed when the informer is initialized
+func (ldi *LazyDemandInformer) Ready() <-chan struct{} {
+	return ldi.ready
+}
+
+// Run starts the goroutine to check for the existence of the demand CRD,
+// and initialize the demand informer if CRD exists
+func (ldi *LazyDemandInformer) Run(ctx context.Context) {
+	if ldi.checkDemandCRDExists(ctx) {
+		return
+	}
+	go func() {
+		_ = wapp.RunWithFatalLogging(ctx, ldi.doStart)
+	}()
+}
+
+func (ldi *LazyDemandInformer) doStart(ctx context.Context) error {
+	t := time.NewTicker(time.Minute)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-t.C:
+			if ldi.checkDemandCRDExists(ctx) {
+				return nil
+			}
+		}
+	}
+}
+
+func (ldi *LazyDemandInformer) checkDemandCRDExists(ctx context.Context) bool {
+	_, ready, err := CheckCRDExists(demandapi.DemandCustomResourceDefinitionName(), ldi.apiExtensionsClient)
+	if err != nil {
+		svc1log.FromContext(ctx).Info("failed to determine if demand CRD exists", svc1log.Stacktrace(err))
+		return false
+	}
+	if ready {
+		svc1log.FromContext(ctx).Info("demand CRD has been initialized. Demand resources can now be created")
+		err = ldi.initializeInformer(ctx)
+		if err != nil {
+			svc1log.FromContext(ctx).Error("failed initializing demand informer", svc1log.Stacktrace(err))
+			return false
+		}
+	}
+	return ready
+}
+
+func (ldi *LazyDemandInformer) initializeInformer(ctx context.Context) error {
+	ldi.lock.Lock()
```

Comment:
I don't think I understand why this lock is needed, can you explain?

Reply:
yea I wanted to block the reader goroutines if there was an ongoing call to this method, but I think it is fine to remove it. Actual guard for the informer field is the ready channel and all readers should wait on the ready channel before reading